-finstrument-functions + libstdc++ + C++17/20 = undefined symbol linker error

|  |  |
| --- | --- |
| Bugzilla Link | [51400](https://llvm.org/bz51400) |
| Version | trunk |
| OS | Linux |
| Attachments | [MRE and test script](https://user-images.githubusercontent.com/93435533/143762403-3b781c0b-f30a-4bed-bffc-f44270f64496.gz) |
| CC | @mattfbacon,@zygoloid |

## Extended Description 


When compiling with `-finstrument-functions` and any standard above C++17, that is C++2a aka 20 or 2b aka 23, when attempting to declare a string, the proper templates relating to standard library allocators are not included in the object files, leading to the following linker error:

```
/usr/bin/ld: /tmp/main-6f2635.o: in function `std::allocator_traits<std::allocator<char> >::deallocate(std::allocator<char>&, char*, unsigned long)':
main.cpp:(.text._ZNSt16allocator_traitsISaIcEE10deallocateERS0_Pcm[_ZNSt16allocator_traitsISaIcEE10deallocateERS0_Pcm]+0x5d): undefined reference to `std::allocator<char>::deallocate(char*, unsigned long)'
/usr/bin/ld: main.cpp:(.text._ZNSt16allocator_traitsISaIcEE10deallocateERS0_Pcm[_ZNSt16allocator_traitsISaIcEE10deallocateERS0_Pcm]+0x99): undefined reference to `std::allocator<char>::deallocate(char*, unsigned long)'
clang-14: error: linker command failed with exit code 1 (use -v to see invocation)
```

Additionally, if the string has contents, the error will include the inability to find an allocate method:

```
/usr/bin/ld: /tmp/main-248bd2.o: in function `std::allocator_traits<std::allocator<char> >::deallocate(std::allocator<char>&, char*, unsigned long)':
main.cpp:(.text._ZNSt16allocator_traitsISaIcEE10deallocateERS0_Pcm[_ZNSt16allocator_traitsISaIcEE10deallocateERS0_Pcm]+0x5d): undefined reference to `std::allocator<char>::deallocate(char*, unsigned long)'
/usr/bin/ld: main.cpp:(.text._ZNSt16allocator_traitsISaIcEE10deallocateERS0_Pcm[_ZNSt16allocator_traitsISaIcEE10deallocateERS0_Pcm]+0x99): undefined reference to `std::allocator<char>::deallocate(char*, unsigned long)'
/usr/bin/ld: /tmp/main-248bd2.o: in function `std::allocator_traits<std::allocator<char> >::allocate(std::allocator<char>&, unsigned long)':
main.cpp:(.text._ZNSt16allocator_traitsISaIcEE8allocateERS0_m[_ZNSt16allocator_traitsISaIcEE8allocateERS0_m]+0x49): undefined reference to `std::allocator<char>::allocate(unsigned long)'
/usr/bin/ld: main.cpp:(.text._ZNSt16allocator_traitsISaIcEE8allocateERS0_m[_ZNSt16allocator_traitsISaIcEE8allocateERS0_m]+0x81): undefined reference to `std::allocator<char>::allocate(unsigned long)'
clang-14: error: linker command failed with exit code 1 (use -v to see invocation)
```

This manifests for `-std=c++20`, `-std=gnu++20`, `-std=c++2b`, `-std=gnu++2b`, but not for `std=c++17`. Additionally, the g++ can compile the code with the same flags, for every C++ standard I tested, from c++17 to c++2b.

See attached tarball for a minimal reproducible example and a script to test whether clang++ and g++ can build it for each standard.

I wasn't sure if this qualifies as release blocker because theoretically you could remove `-finstrument-functions`, in which case it compiles just fine. However I like seeing all the function names in debugging tools like ASAN.

--- a/llvm/lib/Transforms/Utils/EntryExitInstrumenter.cpp
+++ b/llvm/lib/Transforms/Utils/EntryExitInstrumenter.cpp
@@ -103,6 +103,12 @@ static bool runOnFunction(Function &F, bool PostInlining) {
   if (F.hasFnAttribute(Attribute::Naked))
     return false;
 
+  // available_externally functions may not have definitions external to the
+  // module (e.g. gnu::always_inline). Instrumenting them would lead to linker
+  // errors. Skip them like GCC.
+  if (F.hasAvailableExternallyLinkage())
+    return false;
+
   StringRef EntryAttr = PostInlining ? "instrument-function-entry-inlined"
                                      : "instrument-function-entry";
 

--- a/llvm/lib/Transforms/Utils/EntryExitInstrumenter.cpp
+++ b/llvm/lib/Transforms/Utils/EntryExitInstrumenter.cpp
@@ -104,8 +104,8 @@ static bool runOnFunction(Function &F, bool PostInlining) {
     return false;
 
   // available_externally functions may not have definitions external to the
-  // module (e.g. gnu::always_inline). Instrumenting them would lead to linker
-  // errors. Skip them like GCC.
+  // module (e.g. gnu::always_inline). Instrumenting them might lead to linker
+  // errors if they are optimized out. Skip them like GCC.
   if (F.hasAvailableExternallyLinkage())
     return false;
 

--- a/llvm/test/Transforms/EntryExitInstrumenter/mcount.ll
+++ b/llvm/test/Transforms/EntryExitInstrumenter/mcount.ll
@@ -129,6 +129,13 @@ define void @naked() naked {
   ret void
 }
 
+define available_externally void @always_inline() {
+; CHECK-LABEL: define available_externally void @always_inline() {
+; CHECK-NEXT:    ret void
+;
+  ret void
+}
+
 ; The attributes are "consumed" when the instrumentation is inserted.
 ; CHECK: attributes
 ; CHECK-NOT: instrument-function

